src/core/sock.c

```diff
@@ -242,9 +243,51 @@ int nn_sock_term (struct nn_sock *self)
     /*  At this point, we can be reasonably certain that no other thread
         has any references to the socket. */
 
+    nn_sock_unsafe_cleanup(self, NN_CLEAN_DEFAULT);
+
+    return 0;
+}
+
+void nn_release_endpoint (struct nn_list_item *it)
+{
+    struct nn_ep *ep = nn_cont (it, struct nn_ep, item);
+    /* TODO: force cleanup the endpoint */
+    nn_free (ep);
+}
+
+void nn_release_stopped_endpoint (struct nn_list_item *it)
+{
+    struct nn_ep *ep = nn_cont (it, struct nn_ep, item);
+    nn_free (ep);
+}
+
+void nn_sock_unsafe_cleanup (struct nn_sock *self, enum nn_cleanup_opt opts)
+{
+    int i;
+
+    /*  Cleanup. */
+    if (opts & NN_CLEAN_EMPTY) {
+        /* TODO: Force destroy the socket base */
+        /* self->sockbase->vfptr->destroy (self->sockbase); */
```

Comment:
Same as the endpoints, this needs a hook inside the socket base to release everything. The destroy function simply fails because it checks that the state is valid, which cannot happen during a fork, before terminating.